README.md
# newbcsunday README

This is the README for your extension "newbcsunday". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "newbcsunday" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

media/main.js
console.log("log test")
const vscode = acquireVsCodeApi(); // Get the vscode API object

const submitBtnElement = document.getElementById('submitBtn');
const addPort = document.getElementById('addPort');

let enteredData = ''; // Initialize a variable to store user input

submitBtnElement.addEventListener('click', () => {
    console.log("clicked");
    const value = addPort.value;
    enteredData = value; // Store the user input
    vscode.postMessage({ command: 'addPort', value }); // Send the input to extension
});

document.addEventListener('click', (event) => {
    if (event.target && event.target.tagName === 'BUTTON' && event.target.textContent === 'X') {
      const listItem = event.target.closest('.list-item-container');
      if (listItem) {
        const port = listItem.querySelector('.port-name').textContent;
        removePort(port);
      }
    }
  });

  function removePort(port) {
    const message = {
      command: 'removePort',
      value: port,
    };
    vscode.postMessage(message);
  }

  function handleLinkClick(event, key, value) {
    event.preventDefault(); // Prevent the default behavior of opening the link
    vscode.postMessage({command:'handleLinkClick',key,value});
}

out/bcTunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onAllPorts = exports.onUrlSetError = exports.onUrlSet = exports.removePort = exports.getAllPorts = exports.addPort = void 0;
const socket_io_1 = require("socket.io");
let interval;
const io = new socket_io_1.Server({ /* options */});
let tmoleSocket;
function addPort(port) {
    tmoleSocket.emit("addPort", { "port": port });
}
exports.addPort = addPort;
function getAllPorts() {
    tmoleSocket.emit("getAllPorts", null);
}
exports.getAllPorts = getAllPorts;
function removePort(port) {
    tmoleSocket.emit("removePort", { "port": port });
}
exports.removePort = removePort;
let onUrlSetFunc;
let onUrlSetErrorFunc;
let onGetAllPortsFunc;
//function 
function onUrlSet(func) {
    onUrlSetFunc = func;
}
exports.onUrlSet = onUrlSet;
function onUrlSetError(func) {
    onUrlSetErrorFunc = func;
}
exports.onUrlSetError = onUrlSetError;
function onAllPorts(func) {
    onGetAllPortsFunc = func;
}
exports.onAllPorts = onAllPorts;
io.on("connection", (socket) => {
    // ...
    tmoleSocket = socket;
    tmoleSocket.on("urlSet", (data) => {
        if (data.httpUrl) {
            console.log("urlSet", data);
            onUrlSetFunc(data);
        }
    });
    tmoleSocket.on("urlSetError", (data) => {
        onUrlSetErrorFunc(data);
        switch (data.type) {
            case "urlFetchFailed":
                console.log("reason ", data.reason);
                break;
        }
    });
    tmoleSocket.on("allPorts", (data) => {
        console.log("All ports", data);
        onGetAllPortsFunc(data);
    });
    tmoleSocket.on("error", (data) => {
        console.log("ERROR");
    });
    interval = setInterval(() => {
        if (tmoleSocket.connected)
            tmoleSocket.emit("ping", "");
        else
            clearInterval(interval);
    }, 3000);
    getAllPorts();
});
io.listen(101);
//# sourceMappingURL=bcTunnel.js.map

src/bcTunnel.ts

import { Server,Socket } from "socket.io";
interface IUrlSet{
    port:Number,
    httpUrl:string
}

interface IUrlSetError{
    type:string,
    port:Number,
    reason:string
}

interface IUrlGetAll{
    ports:IUrlSet[]
}

let interval:NodeJS.Timer
const io = new Server({ /* options */ });
let tmoleSocket:Socket;
function addPort(port:Number){
    tmoleSocket.emit("addPort", { "port": port})

}

function getAllPorts()
{
    tmoleSocket.emit("getAllPorts", null);
}

function removePort(port:Number)
{
    tmoleSocket.emit("removePort", { "port": port})
}

type onUrlSetCallback= (data:IUrlSet) => void
let onUrlSetFunc:onUrlSetCallback

type onUrlSetErrorCallback=(data:IUrlSetError) => void
let onUrlSetErrorFunc:onUrlSetErrorCallback

type onAllPortsCallback=(data:IUrlGetAll)=> void
let onGetAllPortsFunc:onAllPortsCallback



//function 
 function onUrlSet(func:onUrlSetCallback)
 {
    onUrlSetFunc=func
 }

 function onUrlSetError(func:onUrlSetErrorCallback)
 {
    onUrlSetErrorFunc=func
 }

 function onAllPorts(func:onAllPortsCallback)
 {
    onGetAllPortsFunc=func
 }

 
io.on("connection", (socket) => {
    // ...
    tmoleSocket = socket;
    tmoleSocket.on("urlSet", (data:IUrlSet) => {
        if (data.httpUrl) {
            console.log("urlSet", data);
            onUrlSetFunc(data);
        }
    });

    tmoleSocket.on("urlSetError", (data:IUrlSetError) => {
        onUrlSetErrorFunc(data)
        switch (data.type) {
            case "urlFetchFailed":
                console.log("reason ", data.reason);
                break;
        }
    });

    tmoleSocket.on("allPorts", (data:IUrlGetAll) => {
        console.log("All ports", data);
        onGetAllPortsFunc(data)
    });

    tmoleSocket.on("error",(data)=>{
        console.log("ERROR");
    });
    

    interval = setInterval(() => {
        if (tmoleSocket.connected)
            tmoleSocket.emit("ping", "");
        else
            clearInterval(interval);
    }, 3000);
    getAllPorts()
});


io.listen(101);
export {addPort,getAllPorts,removePort,onUrlSet,onUrlSetError,onAllPorts}

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
const vscode = require("vscode");
const SidebarProvider_1 = require("./SidebarProvider");
// @ts-ignore
const ConfigParser = require("@webantic/nginx-config-parser");
function activate(context) {
    const sidebarProvider = new SidebarProvider_1.SidebarProvider(context.extensionUri);
    context.subscriptions.push(vscode.window.registerWebviewViewProvider("bcportexp-sidebar", sidebarProvider));
    vscode.commands.registerCommand('newbcsunday.helloWorld', () => {
        // Read and parse the .bcport file
        ///  const filePath = path.join(context.extensionPath, 'dummy.bcport');
        //const fileContent = fs.readFileSync(filePath, 'utf-8');
        //const keyValuePairs = fileContent.split('\n').map(line => line.trim());
    });
    // console.log(`${}`);
    var parser = new ConfigParser();
    console.log(`${parser}`);
    // parse straight from file. by default, will try to resolve includes
    var config = parser.readConfigFile('C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf');
    let configNew = parser.parse(config);
    console.log(`${configNew}`);
    vscode.window.showInformationMessage(`${configNew}`);
    // to keep deterministic behaviour, set parseIncludes = false in the options
    // var configWithoutIncludes = parser.readConfigFile('/C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf', { parseIncludes: false })
    // console.log(`${configWithoutIncludes}`);
    // write direct to file (overwriting existing one)
    // parser.writeConfigFile('/C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf', config, true)
    var sampleConfig = {
        "server": {
            "server_name": "_",
            "location /": {
                "try_files": "*.html"
            }
        }
    };
    // to multi-line config string
    var configString = parser.toConf(sampleConfig);
    // and back again
    var configJson = parser.toJSON(configString);
    // shorthand (will change object --> string and string --> object)
    parser.parse(configString);
}
exports.activate = activate;
// Activating extension 'undefined_publisher.newbcsunday' failed: Cannot find module '@webantic/nginx-config-parser' Require stack: - c:\Users\Vishwanath B Hiremat\Downloads\Accenture for 2024 Batch\Accenture 2023\newbcport\newbcsunday\out\extension.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\loader.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-amd.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js - .
//server { listen 4000; listen [::]:4000; location /static/ { proxy_pass http://127.0.0.1:3000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header Host $host; proxy_cache_bypass $http_upgrade; } location / { proxy_pass http://127.0.0.1:3000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header Host $host; proxy_cache_bypass $http_upgrade; } location /listen_6000/ { proxy_pass http://127.0.0.1:6000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header ..
//# sourceMappingURL=extension.js.map

out/SidebarProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SidebarProvider = void 0;
const vscode = require("vscode");
const bcTunnel_1 = require("./bcTunnel");
function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}
let portMapping = new Map();
let portErrorTable = new Map();
//javascript .map function -> takes a list -> iterates list.map() , 
//let portUrlMap = new Map()
//portUrlMap.set() +1,.....n
//portUrlMap.map((kv) => {return ( <li> kv.key , kv.valu <button> X</button> </li>)})
class SidebarProvider {
    constructor(_extensionUri) {
        this._extensionUri = _extensionUri;
        this.isAssigning = false;
        this.addPortBtnText = 'Add Port';
        this.addPortBtnState = '';
        setTimeout(() => { (0, bcTunnel_1.getAllPorts)(); }, 2000);
    }
    resolveWebviewView(webviewView) {
        this._view = webviewView;
        let flag = true;
        //port isAssigning port false
        //create enum{200:success,400:server error,111:processing}
        (0, bcTunnel_1.onAllPorts)((data) => {
            flag = true;
            for (let i = 0; i < data.ports.length; i++) {
                let current = data.ports[i];
                portMapping.set(current.port, current.httpUrl);
            }
            this.updateWebviewContent(webviewView.webview, true); //here flag should be true
        });
        (0, bcTunnel_1.onUrlSet)((data) => {
            portMapping.set(data.port, data.httpUrl);
            //isAssigning port is false
            this.isAssigning = false;
            vscode.window.showInformationMessage(`Port Mapping Success: ${data.port}`);
            this.updateWebviewContent(webviewView.webview, true);
        });
        (0, bcTunnel_1.onUrlSetError)((data) => {
            portMapping.set(data.port, "x");
            portErrorTable.set(data.port, data.type);
            vscode.window.showErrorMessage(`Port Mapping Failure: ${data.port},Reason:${data.reason}`);
            this.isAssigning = false;
            this.updateWebviewContent(webviewView.webview, true);
        });
        webviewView.webview.options = {
            // Allow scripts in the webview
            enableScripts: true,
            localResourceRoots: [this._extensionUri],
        };
        // webviewView.webview.html = this.getWebviewContent(webviewView.webview);
        webviewView.webview.onDidReceiveMessage(async (message) => {
            switch (message.command) {
                //change userInput to addPort and storedInput to portToAdd--------Done
                //if(portMapping.has(portToAdd) then show warning message)--------Done
                //try and learn httpget and httppost in JavaScript
                //icon------------------------------------------------------------Done
                case 'addPort': {
                    //portMapping.set(storedInput,"12123");
                    let portToAdd = message.value; // Store the received user input, value comes in message.value
                    if (portMapping.has(portToAdd)) {
                        console.log(' Warning Message sent to extension that port is already added:', portToAdd); // Log the message
                        vscode.window.showWarningMessage(`Warning:Port already added ${portToAdd}`);
                        break;
                    }
                    this.isAssigning = true;
                    (0, bcTunnel_1.addPort)(portToAdd);
                    //isAssigning port=true
                    this.isAssigning = true;
                    this.updateWebviewContent(webviewView.webview, true);
                    break;
                }
                case 'removePort': {
                    const portToRemove = Number(message.value);
                    /*   if(!portMapping.has(portToRemove))
                       {
                          console.log('Warning-Port not found:',portToRemove);
                          vscode.window.showMessage(`Port not found:${portToRemove}`);
                       }*/
                    (0, bcTunnel_1.removePort)(portToRemove);
                    portMapping.delete(portToRemove);
                    vscode.window.showInformationMessage(`Port Removed:${portToRemove}`);
                    this.updateWebviewContent(webviewView.webview, true);
                    break;
                }
                case 'handleLinkClick': {
                    const { key, value } = message;
                    try {
                        vscode.window.showErrorMessage(`Error while handling link: ${key}`);
                    }
                    catch (error) {
                        vscode.window.showErrorMessage(`Error while handling link: ${error}`);
                    }
                    break;
                }
            }
        });
        //webviewView.webview.html = this.getWebviewContent(webviewView.webview);
        this.updateWebviewContent(webviewView.webview, false); //here flag should be false
    }
    //flag that tells loaded or not--->put this flag in updateContentView parameters
    updateWebviewContent(webview, loadState) {
        let newHtml = '';
        if (loadState) {
            newHtml = this.getWebviewContent(webview);
        }
        //`<p> Loading ... >/p>`
        else {
            newHtml = `<p> Loading...</p>`;
        }
        webview.html = newHtml;
    }
    revive(panel) {
        this._view = panel;
    }
    generateListItems() {
        let listItems = '';
        for (const [key, value] of portMapping.entries()) {
            listItems += ` <li>
        <div class="list-item-container">
        <div>${key}:${value === "x" ?
                `<p>${portErrorTable.get(key)}</p>`
                :
                    `<a href="${value}" onclick="handleLinkClick(event, '${key}', '${value}')">${value}</a>`}
          </div>
        <button onClick="removePort('${key}')" style="min-width:25px;width:25px;"> X</button>
        </div>
      </li>`;
            console.log(listItems);
        }
        return listItems;
    }
    getWebviewContent(webview) {
        /* const styleResetUri = webview.asWebviewUri(
           vscode.Uri.joinPath(this._extensionUri, "media", "reset.css")
         );*/
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js'));
        const styleCssPath = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'styles.css'));
        const vsCssPath = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'vscode.css'));
        /*const styleMainUri = webview.asWebviewUri(
          vscode.Uri.joinPath(this._extensionUri, "out", "compiled/sidebar.css")
        );*/
        const listItems = this.generateListItems();
        let addPortBtnHtml = '';
        this.addPortBtnText = this.isAssigning ? 'Assigning Port...' : 'Add Port';
        this.addPortBtnState = this.isAssigning ? 'disabled' : '';
        addPortBtnHtml += `
    <div class="list-item-container">
        <button id="submitBtn" ${this.addPortBtnState}>${this.addPortBtnText}</button>
        <input type="text" id="addPort" ${this.addPortBtnState} placeholder="Enter a value"/>
    </div>
`;
        // Use a nonce to only allow a specific script to be run.
        const nonce = getNonce();
        return `
    <!DOCTYPE html>
    <html>
    <head>
        <style>
            body { font-family: Arial, sans-serif; }
            .list-item-container {
                display: flex;
                align-items: center; /* Center items vertically */
                justify-content: space-between; 
                margin-bottom: 10px;
              }

              .list-item-container button {
                margin-left: 10px; 
              }

        </style>
        <link href="${styleCssPath}" rel="stylesheet">
        <link href="${vsCssPath}" rel="stylesheet">
    </head>
    <body>
        <h3>Exposed Ports</h3>

        <ul> 
        

       ${listItems}
        
        </ul>
       ${addPortBtnHtml}

        <script nonce="${nonce}" src="${scriptUri}">
      
        </script>
    </body>
    </html>
`;
    }
}
exports.SidebarProvider = SidebarProvider;
//# sourceMappingURL=SidebarProvider.js.map

src/SidebarProvider.ts
import * as vscode from 'vscode';
import {addPort,getAllPorts,onAllPorts,onUrlSet,onUrlSetError,removePort} from './bcTunnel';
function getNonce() {
	let text = '';
	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < 32; i++) {
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	}
	return text;
}


let portMapping = new Map();
let portErrorTable=new Map();
//javascript .map function -> takes a list -> iterates list.map() , 
//let portUrlMap = new Map()
//portUrlMap.set() +1,.....n
//portUrlMap.map((kv) => {return ( <li> kv.key , kv.valu <button> X</button> </li>)})
export class SidebarProvider implements vscode.WebviewViewProvider {
  _view?: vscode.WebviewView;
  _doc?: vscode.TextDocument;

  private isAssigning:boolean =false;
  private addPortBtnText='Add Port'
  private addPortBtnState=''
  constructor(private readonly _extensionUri: vscode.Uri) {
   setTimeout(()=>{getAllPorts()},2000)
  }
 
  
  public resolveWebviewView(webviewView: vscode.WebviewView) {
    this._view = webviewView;
    let flag:boolean=true;
    //port isAssigning port false
    //create enum{200:success,400:server error,111:processing}

    onAllPorts((data)=>{
      flag=true;
      for(let i=0;i<data.ports.length;i++)
      {
        let current=data.ports[i]
        portMapping.set(current.port,current.httpUrl) 
      }
      this.updateWebviewContent(webviewView.webview,true);//here flag should be true
    })

    onUrlSet((data)=>{
      portMapping.set(data.port,data.httpUrl);
      //isAssigning port is false
      this.isAssigning=false;
      vscode.window.showInformationMessage(`Port Mapping Success: ${data.port}`);
      this.updateWebviewContent(webviewView.webview,true);
     })

     onUrlSetError((data)=>{
      portMapping.set(data.port,"x");
      portErrorTable.set(data.port,data.type);
      vscode.window.showErrorMessage(`Port Mapping Failure: ${data.port},Reason:${data.reason}`);
      this.isAssigning=false;
    this.updateWebviewContent(webviewView.webview,true);

     })
    webviewView.webview.options = {
      // Allow scripts in the webview
      enableScripts: true,

      localResourceRoots: [this._extensionUri],
    };

   // webviewView.webview.html = this.getWebviewContent(webviewView.webview);

    webviewView.webview.onDidReceiveMessage(async (message) => {
      switch (message.command) {
          //change userInput to addPort and storedInput to portToAdd--------Done
          //if(portMapping.has(portToAdd) then show warning message)--------Done
          //try and learn httpget and httppost in JavaScript
          //icon------------------------------------------------------------Done
        case 'addPort' :{
            //portMapping.set(storedInput,"12123");
           let portToAdd = message.value; // Store the received user input, value comes in message.value
           if(portMapping.has(portToAdd))
           {
            console.log(' Warning Message sent to extension that port is already added:', portToAdd); // Log the message
            vscode.window.showWarningMessage(`Warning:Port already added ${portToAdd}`);
            break;
           }
           this.isAssigning=true;

         addPort(portToAdd)
        //isAssigning port=true
        this.isAssigning=true;

        this.updateWebviewContent(webviewView.webview,true);
        break;
        }

        case 'removePort':{
            const portToRemove = Number(message.value);
         /*   if(!portMapping.has(portToRemove))
            {
               console.log('Warning-Port not found:',portToRemove);
               vscode.window.showMessage(`Port not found:${portToRemove}`);
            }*/
            removePort(portToRemove)
            portMapping.delete(portToRemove);
            vscode.window.showInformationMessage(`Port Removed:${portToRemove}`);
            
            this.updateWebviewContent(webviewView.webview,true);
            break;
        }

        case 'handleLinkClick':{
          const { key, value } = message;
            try {
                vscode.window.showErrorMessage(`Error while handling link: ${key}`);

            } catch (error) {
                vscode.window.showErrorMessage(`Error while handling link: ${error}`);
            }
            break;
        }
      }
    });
    //webviewView.webview.html = this.getWebviewContent(webviewView.webview);
    this.updateWebviewContent(webviewView.webview,false);//here flag should be false

  }
//flag that tells loaded or not--->put this flag in updateContentView parameters
  private updateWebviewContent(webview: vscode.Webview,loadState:boolean): void {
    let newHtml='';
    if(loadState)
    {
       newHtml = this.getWebviewContent(webview);
    }
    //`<p> Loading ... >/p>`
    else{
       newHtml=`<p> Loading...</p>`
    }
    webview.html = newHtml;
}

  public revive(panel: vscode.WebviewView) {
    this._view = panel;
  } 

  

  private generateListItems(): string {
    let listItems = '';
    for (const [key, value] of portMapping.entries()) {
        listItems += ` <li>
        <div class="list-item-container">
        <div>${key}:${value==="x" ?
        `<p>${portErrorTable.get(key)}</p>`
        :
        `<a href="${value}" onclick="handleLinkClick(event, '${key}', '${value}')">${value}</a>`
      
      }
          </div>
        <button onClick="removePort('${key}')" style="min-width:25px;width:25px;"> X</button>
        </div>
      </li>`;
        console.log(listItems);
    }
    return listItems;
}


  

  private getWebviewContent(webview: vscode.Webview) {
   /* const styleResetUri = webview.asWebviewUri(
      vscode.Uri.joinPath(this._extensionUri, "media", "reset.css")
    );*/
    const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js'));
    const styleCssPath= webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri,'media','styles.css'));
    const vsCssPath= webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri,'media','vscode.css'));
    /*const styleMainUri = webview.asWebviewUri(
      vscode.Uri.joinPath(this._extensionUri, "out", "compiled/sidebar.css")
    );*/
 
    const listItems = this.generateListItems();

    let addPortBtnHtml=''

    this.addPortBtnText = this.isAssigning ? 'Assigning Port...' : 'Add Port';
    this.addPortBtnState=this.isAssigning? 'disabled':'';

    addPortBtnHtml += `
    <div class="list-item-container">
        <button id="submitBtn" ${this.addPortBtnState}>${this.addPortBtnText}</button>
        <input type="text" id="addPort" ${this.addPortBtnState} placeholder="Enter a value"/>
    </div>
`;
    // Use a nonce to only allow a specific script to be run.
    const nonce = getNonce();


    return `
    <!DOCTYPE html>
    <html>
    <head>
        <style>
            body { font-family: Arial, sans-serif; }
            .list-item-container {
                display: flex;
                align-items: center; /* Center items vertically */
                justify-content: space-between; 
                margin-bottom: 10px;
              }

              .list-item-container button {
                margin-left: 10px; 
              }

        </style>
        <link href="${styleCssPath}" rel="stylesheet">
        <link href="${vsCssPath}" rel="stylesheet">
    </head>
    <body>
        <h3>Exposed Ports</h3>

        <ul> 
        

       ${listItems}
        
        </ul>
       ${addPortBtnHtml}

        <script nonce="${nonce}" src="${scriptUri}">
      
        </script>
    </body>
    </html>
`;
  }
}

src/extension.ts

import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import { SidebarProvider } from './SidebarProvider';
 // @ts-ignore
import  * as ConfigParser from '@webantic/nginx-config-parser'

export function activate(context: vscode.ExtensionContext) {

    const sidebarProvider = new SidebarProvider(context.extensionUri);
    context.subscriptions.push(
        vscode.window.registerWebviewViewProvider("bcportexp-sidebar", sidebarProvider)
      );

    vscode.commands.registerCommand('newbcsunday.helloWorld', () => {
        // Read and parse the .bcport file
      ///  const filePath = path.join(context.extensionPath, 'dummy.bcport');
        //const fileContent = fs.readFileSync(filePath, 'utf-8');
        //const keyValuePairs = fileContent.split('\n').map(line => line.trim());

    });

    // console.log(`${}`);
var parser =new ConfigParser()
console.log(`${parser}`);
// parse straight from file. by default, will try to resolve includes
var config = parser.readConfigFile('C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf');
let configNew = parser.parse(config)
console.log(`${configNew}`);
vscode.window.showInformationMessage(`${configNew}`);
// to keep deterministic behaviour, set parseIncludes = false in the options
// var configWithoutIncludes = parser.readConfigFile('/C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf', { parseIncludes: false })
// console.log(`${configWithoutIncludes}`);
// write direct to file (overwriting existing one)
// parser.writeConfigFile('/C:\\Users\\Vishwanath B Hiremat\\Downloads\\Accenture for 2024 Batch\\Accenture 2023\\newbcport\\newbcsunday\\example.conf', config, true)


var sampleConfig = {
  "server": {
    "server_name": "_",
    "location /": {
      "try_files": "*.html"
    }
  }
}

// to multi-line config string
var configString = parser.toConf(sampleConfig)
// and back again
var configJson = parser.toJSON(configString)

// shorthand (will change object --> string and string --> object)
parser.parse(configString)
}
// Activating extension 'undefined_publisher.newbcsunday' failed: Cannot find module '@webantic/nginx-config-parser' Require stack: - c:\Users\Vishwanath B Hiremat\Downloads\Accenture for 2024 Batch\Accenture 2023\newbcport\newbcsunday\out\extension.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\loader.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-amd.js - c:\Users\Vishwanath B Hiremat\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js - .
//server { listen 4000; listen [::]:4000; location /static/ { proxy_pass http://127.0.0.1:3000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header Host $host; proxy_cache_bypass $http_upgrade; } location / { proxy_pass http://127.0.0.1:3000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header Host $host; proxy_cache_bypass $http_upgrade; } location /listen_6000/ { proxy_pass http://127.0.0.1:6000; proxy_http_version 1.1; proxy_set_header Upgrade $http_upgrade; proxy_set_header Connection 'upgrade'; proxy_set_header ..
